stdlib/src/binascii.rs

```diff
@@ -171,25 +177,107 @@ mod decl {
+            }
+
+            if strict_mode && b[0] == PAD {
+                return Err(base64::DecodeError::InvalidByte(0, 61));
+            }
+
+            let mut decoded: Vec<u8> = vec![];
+
+            let mut quad_pos = 0; // position in the nibble
+            let mut pads = 0;
+            let mut left_char: u8 = 0;
+            let mut padding_started = false;
+            for (i, &el) in b.iter().enumerate() {
+                if el == PAD {
+                    padding_started = true;
+
+                    pads += 1;
+                    if quad_pos >= 2 && quad_pos + pads >= 4 {
+                        if strict_mode && i + 1 < b.len() {
+                            // Represents excess data after padding error
+                            return Err(base64::DecodeError::InvalidLastSymbol(i, PAD));
+                        }
+
+                        return Ok(decoded);
+                    }
+
+                    continue;
                 }
-                decode(&buf)
-            })
+
+                let binary_char = BASE64_TABLE[el as usize];
+                if binary_char >= 64 || binary_char == -1 {
+                    if strict_mode {
+                        // Represents non-base64 data error
+                        return Err(base64::DecodeError::InvalidByte(i, el));
+                    }
+                    continue;
+                }
+
+                if strict_mode && padding_started {
+                    // Represents discontinuous padding error
+                    return Err(base64::DecodeError::InvalidByte(i, PAD));
+                }
+                pads = 0;
+
+                // Decode individual ASCII character
+                if quad_pos == 0 {
+                    quad_pos = 1;
+                    left_char = binary_char as u8;
+                } else if quad_pos == 1 {
+                    quad_pos = 2;
+                    decoded.push((left_char << 2) | (binary_char >> 4) as u8);
+                    left_char = (binary_char & 0x0f) as u8;
+                } else if quad_pos == 2 {
+                    quad_pos = 3;
+                    decoded.push((left_char << 4) | (binary_char >> 2) as u8);
+                    left_char = (binary_char & 0x03) as u8;
+                } else if quad_pos == 3 {
+                    quad_pos = 0;
+                    decoded.push((left_char << 6) | binary_char as u8);
+                    left_char = 0;
+                }
+            }
+
+            if quad_pos == 1 {
+                // Ensure that a PAD never gets passed, since that'd mistakenly cause an excess
+                // data after padding error
+                return Err(base64::DecodeError::InvalidLastSymbol(decoded.len() / 3 * 4 + 1, 0));
+            } else if quad_pos > 1 {
+                return Err(base64::DecodeError::InvalidLength)
+            }
+
+            Ok(decoded)
+        })
+        .map_err(|err| {
+            let python_error = match err {
```

Comment:
`impl ToPyException for base64::DecodeError` is possible to replace this block

Reply:
Sorry, pretty new to Rust. Could you please provide an example? Not super sure what you mean :grimacing:

Reply:
`io.rs` in `vm/src/stdlib` contains an implementation of `ToPyException` for `Io::Error`which can definitely serve as a good example. You basically just move the `match` in there and use `to_pyexception` here.

Reply:
We have a trait called `ToPyException` which used to convert any error to python exception. It is used to convert `Io::Error` to python exception as @DimitrisJim explained.
Once you implement it, this code will look like:

```rust
        .map_err(|err| err.to_pyexception(vm))
```

If you stuck with it, please feel free to ask anything again or tell us to take over that part.

The patch is really nice and this suggestion is totally optional. You don't need to finish to implement ToPyException to finish this PR.

Thank you!

Reply:
Thanks for the kind words! I'm really sorry - it looks like work is getting pretty busy, so I'm not sure that I'll be able to implement this. Possible for you to take it on? @youknowone